### ISSUE TITLE
Update build configuration to publish to JCenter and Maven Central

### ISSUE THREAD

**Issue by alexanderdean** [snowplow/snowplow-scala-tracker#31]

**Comment by chuwy** [comment]
Hey @ninjabear! Could you provide me example so I could implement it (and [this](https://github.com/snowplow/snowplow-scala-tracker/issues/32) one) by myself?

**Comment by ninjabear** [comment]
Hey @chuwy! The sample setup is here https://github.com/snowplow/scala-util/tree/feature/sample-ci-cd, however I'd hold fire for now - there's a couple of things to do before we can "go live" with CD (on any project) e.g. sort out the proper API keys and making sure it lands in the right package (on bintray). 

I hope to get back on this tomorrow - I'll keep you posted!

**Comment by chuwy** [comment]
@alexanderdean may be we can release without CD stuff then? I also started two parallel branches with:
- Incorporating scala-iglu-core into scala-tracker and 
- Small refactoring that can allow us to inject Akka system, which in turn will allow us to not depend on particular akka/spray versions and even to not depenend on akka at all (using finagle for example). This all true If we're okay to provide additional artifacts (scala-tracker-akka, scala-tracker-finage). Also it can close: #9, #16, #22.

These two features seems a bit experimental and I don't want to put them into 0.3.0. But what I'm trying to say is that next 0.4.0 scala-tracker release can be close as well, so we can put #31 and #32 to it.

**Comment by alexanderdean** [comment]
Hi @chuwy - sure, let's push back the CD side of things...

**Comment by alexanderdean** [comment]
This is working with scala-util, so I think we are good to include this in 0.3.0

**Comment by ninjabear** [comment]
Sorry for the delay @chuwy ! Here's the steps we need to take in order to automatically deploy this project to snowplow-maven (our bintray repo), then on to jcenter & maven central.

# 1. Add the credentials to travis.yml

If it's not done already - @alexanderdean will need to do this. This is so that travis can push to bintray on our behalf. For more info - https://docs.travis-ci.com/user/environment-variables/#Encrypted-Variables.

# 2. Alter your build project to include the bintray-sbt plugin
1. Add a file for the plugin, `plugins.sbt` with the following contents `addSbtPlugin("me.lessis" % "bintray-sbt" % "0.3.0")` 
   For reference - [here's the one for scala-util](https://github.com/snowplow/scala-util/blob/master/project/plugins.sbt).
2. Add the bintray specific settings to your build settings:
   
   ``` scala
   // bintray publishing settings
   lazy val publishSettings = bintraySettings ++ Seq[Setting[_]](
     licenses += ("Apache-2.0", url("http://www.apache.org/licenses/LICENSE-2.0.html")),
     bintrayOrganization := Some("snowplow"),
     bintrayRepository := "snowplow-maven"
   )
   
   lazy val mavenCentralExtras = Seq[Setting[_]](
     pomIncludeRepository := { x => false },
     homepage := Some(url("http://snowplowanalytics.com")),
     scmInfo := Some(ScmInfo(url("https://github.com/snowplow/scala-util"), "scm:git@github.com:snowplow/scala-util.git")),
     pomExtra := (
       <developers>
         <developer>
           <name>Snowplow Analytics Ltd</name>
           <email>support@snowplowanalytics.com</email>
           <organization>Snowplow Analytics Ltd</organization>
           <organizationUrl>http://snowplowanalytics.com</organizationUrl>
         </developer>
       </developers>)
   )
   
   lazy val buildSettings = basicSettings ++ publishSettings ++ mavenCentralExtras
   ```
   
     NB you'll need to change the `homepage` and `scmInfo` to match this project. A full example is available [here](https://github.com/snowplow/scala-util/blob/master/project/BuildSettings.scala#L31-L52).

At this point, this project can be pushed to bintray using `sbt publish`. The next step is to have travis run this for us [on certain conditions](https://docs.travis-ci.com/user/deployment/#Conditional-Releases-with-on%3A).

# 3. Changes for travis

We're using tags for releases - so when we push a tag, it'll cause travis to build and release the artifacts (from any branch). There's a catch to the `bintray-sbt` plugin that is it's not possible to specify the credentials in the way travis exposes them (as environment variables). Until this issue is fixed we'll need to work around it by creating the file it expects before building:
1. Create a shell script in a new directory called `.travis/` called `deploy.sh`:
   
   ``` bash
   #!/bin/bash
   
   tag_version=$1
   
   mkdir ~/.bintray/
   FILE=$HOME/.bintray/.credentials
   cat <<EOF >$FILE
   realm = Bintray API Realm
   host = api.bintray.com
   user = $BINTRAY_USER
   password = $BINTRAY_API_KEY
   EOF
   
   cd $TRAVIS_BUILD_DIR
   pwd
   
   project_version=$(sbt version -Dsbt.log.noformat=true | perl -ne 'print $1 if /(\d+\.\d+[^\r\n]*)/')
   if [ "${project_version}" == "${tag_version}" ]; then
       sbt publish
   else 
       echo "Tag version '${tag_version}' doesn't match version in scala project ('${project_version}'). Aborting!"
       exit 1
   fi
   ```
   
   This is the file travis will use to do the deployment.
2. Add the following to `.travis.yml`
   
   ``` yaml
   deploy:
   skip_cleanup: true
   provider: script
   script: ./.travis/deploy.sh $TRAVIS_TAG
   on:
     tags: true
   ```

# Summary

In summary - you'll need to modify the build to include the sbt-release plugin. This handles all the bintray specific stuff. Then you'll need to edit the `.travis.yml` file to cause travis to run `sbt publish` for us on a tag push. Due to a bit of a glitch with sbt-bintray, you need to add the bash script to write out the credentials file prior to `sbt publish`.

There's a little bit more to come - what I've described here will get us to jcenter, but not quite to mavencentral. I'll explain more when we come to it.

**Comment by chuwy** [comment]
@alexanderdean I think we need to rephrase title to reflect that it will be published only to JCenter. And also omit "BuildSettings.scala".

**Comment by alexanderdean** [comment]
Go for it @chuwy !

**Comment by chuwy** [comment]
Hey @ninjabear. Could you please review? https://github.com/snowplow/snowplow-scala-tracker/commit/6e8a5bdfd9e019ab066feeb4a98e3c142977ee90
Also @alexanderdean do we need [Local Maven Repository](https://github.com/snowplow/snowplow-scala-tracker/commit/6e8a5bdfd9e019ab066feeb4a98e3c142977ee90#diff-c6575342216356c040d3b53a22f877c4L49)? I don't quite understand what it does.

**Comment by alexanderdean** [comment]
Hey @chuwy - no you can safely remove the Local Maven Repository. This was part of our old publishing approach (I would basically publish locally and then copy up into our remote Maven).

**Comment by chuwy** [comment]
Good. That is what thought.

**Comment by ninjabear** [comment]
Hey @chuwy, looks good to me! I've added a comment to one of the commits, but they tend to get lost so I'll duplicate here;
- In `.travis/deploy.sh`

``` bash
project_version=$(sbt version -Dsbt.log.noformat=true | perl -ne 'print $1 if /(\d+\.\d+[^\r\n]*)/')
 if [ "${project_version}" == "${tag_version}" ]; then
     sbt publish
 else
```

The `sbt publish` needs to be `sbt +publish` to also publish the artifacts for the cross compiled versions.
- `.travis/deploy.sh` needs to be executable, `chmod +x ..`
- `-SNAPSHOT` needs removing here - https://github.com/snowplow/snowplow-scala-tracker/blob/43c64db9e31a17925646fb66bc14723e687ea6ae/project/BuildSettings.scala#L23

**Comment by alexanderdean** [comment]
Thanks @ninjabear !

**Comment by chuwy** [comment]
Thank you @ninjabear. Updated `deploy.sh`. On `-SNAPSHOT` - do you mean it should be removed on last commit or do you mean it shouldn't appear at all?

**Comment by ninjabear** [comment]
When we push a tag - it'll have to look exactly like the version there. So when we do the final release, yes it'll just be `0.3.0` as this is the version as we want it to appear in jcenter et al. However if you want to push a release candidate or other yourself, `-rcXX` or `-M` is fine, providing the tag still matches up.

The reason for avoiding `SNAPSHOT` specifically is that with tags, we can't replicate what a `SNAPSHOT` is. Artifacts with `-SNAPSHOT` appended in maven land are ongoing releases, so a snapshot is updated over time. If I were to include a dependency that is a snapshot, maven would check periodically based on the timestamp to see if it should re-download it. Since a tag is one point in time, we can't have multiple tags with the same name, which means unless we remove and re-add the tag (not recommended) then we can't really model snapshots. 

It's for these reasons, and the fact that developing in a team that uses snapshots is quite annoying (what version am I actually running?) that bintray doesn't allow them.

**Comment by alexanderdean** [comment]
Great explanation @ninjabear ! We don't use snapshots internally at Snowplow in any case.

**Comment by chuwy** [comment]
Totally makes sense. I'll change it to `M1` on root of this branch then.

**Comment by alexanderdean** [comment]
Can we try pushing the M1 all the way through to MC as a test? Let me know if you need anything from me, otherwise please go ahead

**Comment by chuwy** [comment]
@alexanderdean do you want me to push it with tag `M1`?

**Comment by alexanderdean** [comment]
Yes please

**Comment by chuwy** [comment]
... or `0.3.0-M1`?

**Comment by alexanderdean** [comment]
`0.3.0-M1`
